src/export/ExportWavPack.cpp

```diff
@@ -0,0 +1,430 @@
+   if (!WavpackSetConfiguration64(wpc, &config, -1, NULL) || !WavpackPackInit(wpc)) {
+      WavpackCloseFile(wpc);
+      return ProgressResult::Failed;
+   }
+
+// Samples to write per run
+#define SAMPLES_PER_RUN 8192u
+
+   uint32_t bufferSize = SAMPLES_PER_RUN * numChannels;
+   ArrayOf<int32_t> wavpackBuffer{ bufferSize };
+   auto updateResult = ProgressResult::Success;
+   {
+      auto mixer = CreateMixer(tracks, selectionOnly,
+         t0, t1,
+         numChannels, SAMPLES_PER_RUN, true,
+         rate, int16Sample, mixerSpec);
+
+      InitProgress( pDialog, fName,
+         selectionOnly
+            ? XO("Exporting selected audio as Wavpack file")
+            : XO("Exporting the audio as Wavpack file") );
+      auto &progress = *pDialog;
+
+      while (updateResult == ProgressResult::Success) {
+         auto samplesThisRun = mixer->Process(SAMPLES_PER_RUN);
+
+         if (samplesThisRun == 0)
+            break;
+
+         char *mixed = (char *)(const char*)mixer->GetBuffer();
+         for (decltype(samplesThisRun) j = 0; j < samplesThisRun; j++) {
+            for (size_t i = 0; i < numChannels; i++) {
+               int32_t value = *mixed++ & 0xff;
+               value += *mixed++ << 8;
+               wavpackBuffer[j*numChannels + i] = value;
+            }
+         }
+
+         if (!WavpackPackSamples(wpc, wavpackBuffer.get(), samplesThisRun))
+            updateResult = ProgressResult::Cancelled;
+
+         if (updateResult == ProgressResult::Success)
+            updateResult =
+               progress.Update(mixer->MixGetCurrentTime() - t0, t1 - t0);
+      }
+   }
+
+   if (!WavpackFlushSamples(wpc)) {
+      updateResult = ProgressResult::Cancelled;
+   } else {
+      if (metadata == NULL)
+         metadata = &Tags::Get( *project );
+
+      wxString n;
+      for (const auto &pair : metadata->GetRange()) {
+         n = pair.first;
+         const auto &v = pair.second;
+         if (n == TAG_YEAR) {
+            n = wxT("DATE");
+         }
+         WavpackAppendTagItem(wpc,
+                              (char *) (const char *) n.mb_str(wxConvUTF8),
+                              (char *) (const char *) v.mb_str(wxConvUTF8),
+                              (int) v.length());
+      }
+
+      if (!WavpackWriteTag(wpc)) {
+         // Not sure what to do
+      }
+   }
+
+   WavpackCloseFile(wpc);
+
+   if ( !outWvFile.file.get()->Close()
+      || ( outWvcFile.file && outWvcFile.file.get() && !outWvcFile.file.get()->Close())) {
+      updateResult = ProgressResult::Cancelled;
+   }
+
+   return updateResult;
+}
+
+// Based on the implementation of write_block in dbry/WavPack
+// src: https://github.com/dbry/WavPack/blob/master/cli/wavpack.c
+int ExportWavPack::WriteBlock(void *id, void *data, int32_t length)
+{
+   WriteId *outId = (WriteId *) id;
+
+   uint32_t bcount;
+
+   if (outId->error)
+      return FALSE;
+
+   if (outId && outId->file && outId->file.get() && data && length) {
```

Comment:
This function can be rewritten with less nesting involved!

```
int ExportWavPack::WriteBlock(void *id, void *data, int32_t length)
{
    if (id == nullptr || data == nullptr || length == 0)
        return true; // This is considered to be success in wavpack.c reference code

    WriteId *outId = static_cast<WriteId*>(id);

    if (!outId->file)
        // This does not match the wavpack.c but in our case if file is nullptr - 
        // the stream error has occured
        return false; 

    if (!outId->file->Write(data, length).IsOk()) {
        outId->file.reset();
        return false;
    }

    outId->bytesWritten += length;

    if (outId->firstBlockSize == 0)
        outId->firstBlockSize = length;

    return true;
}
```

Also, this allows to remove `error` from the WriteId structure!